torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+        // and then record the dependency.
+        auto * owning_node = use.user;
+        while (owning_node != nullptr) {
+          auto search = node_to_vertex.find(owning_node);
+          if (search == node_to_vertex.end()) {
+            owning_node = owning_node->owningBlock()->owningNode();
+            continue;
+          }
+          dag.addEdge(node_to_vertex[node], search->second);
+          break;
+        }
+      }
+    }
+  }
+  return dag;
+}
+
+static void find_differentiable_groups(
+    detail::DynamicDAG<Node*>& dep_graph,
+    size_t distance_threshold=64,
+    size_t producer_outedge_threshold=16) {
+  // A Vertex contains a Node* or a differential group of Node*.
+  // Perform graph contraction on dep_graph: contract two vertices(x, y) if
+  // the following conditions hold:
+  // - x, y can be merged to form a differential group
+  // - the contraction would not invalidate the dag (it creates no cycles).
+
+  // Iterate in reverse topological order
+  int64_t ord = dep_graph.maybe_vertices().size() - 1;
+  while (ord >= 0) {
+    if (!dep_graph.maybe_vertices().at(ord)) {
+      --ord;
+      continue;
+    }
+
+    auto * consumer = dep_graph.at(ord);
+    if (!shouldConsiderForMerge(consumer)) {
+      --ord;
+      continue;
+    }
+
+    // Iterate through consumer->in_edges in reverse topological order
+    detail::vertex_list<Node*> in_edges;
+    in_edges.assign(consumer->in_edges.begin(), consumer->in_edges.end());
+    dep_graph.sort(in_edges);
+
+    bool changed = false;
+    for (auto it = in_edges.rbegin(); it != in_edges.rend(); ++it) {
+      auto * producer = *it;
+      // The distance threshold makes this algorithm "not optimal": it will miss
+      // some possible contraction opportunities, but it hopefully lets us:
+      // 1) preserve locality of tensors. We don't want to keep them alive for too long.
+      // 2) Bound the computation complexity for contractEdge
+      if (consumer->ord - producer->ord > distance_threshold) continue;
+      if (producer->out_edges.size() > producer_outedge_threshold) continue;
+      if (!shouldConsiderForMerge(producer)) continue;
+
+      // If the edge contraction is successful, consumer->in_edges
+      // may have changed so we break out of this loop.
+      if (dep_graph.contractEdge(producer, consumer)) {
+        changed = true;
+        break;
+      }
+    }
+    // If we successfully contracted an edge, stay at this vertex.
+    // It may have new edges that should be looked at before moving on.
+    if (!changed) {
+      --ord;
+    }
+  }
+}
+
+static void reorder_according_to_dag(Block * block, const detail::DynamicDAG<Node*>& dep_graph) {
+  auto& vertices = dep_graph.maybe_vertices();
+  for (auto it = vertices.begin(); it != vertices.end(); ++it) {
+    if (!it->has_value()) continue;
+
+    auto& rnodes = it->value()->rdata;
+    for (auto it = rnodes.rbegin(); it != rnodes.rend(); ++it) {
+      (*it)->moveBefore(block->return_node());
```

Comment:
Lol this is a bit surprising at the end, but it makes sense once you realize that we'll move all nodes even if it was unnecessary. A comment might be useful